# **Aula 7 - 22 de setembro de 2026**

## Bibliotecas e Funções

In [62]:
import statistics

In [63]:
def calcular_moda(valores):
    modas = []
    maior_frequencia = 0
    
    for valor in valores:
        frequencia = valores.count(valor)
        
        if frequencia > maior_frequencia:
            maior_frequencia  = frequencia
            modas = [valor]
        
        elif frequencia == maior_frequencia and valor not in modas:
            modas.append(valor)
            
        
    if maior_frequencia == 1:
        return "Amodal"
    
    elif len(modas) == 1:
        return "Unimodal", modas
    
    elif len(modas) == 2:
        return "Bimodal", modas
    
    else:
        return "Multimodal", modas        

In [64]:
def calcula_amplitude(lista):
    maximo = max(lista)
    minimo = min(lista)
    amplitude = maximo - minimo
    return amplitude

In [65]:
def calcula_coeficiente_variacao(lista):
    desvio_padrao = statistics.stdev(lista)
    media = statistics.mean(lista)
    
    coeficiente_variacao = (desvio_padrao / media) * 100
    return coeficiente_variacao

In [66]:
def calcular_quartis(lista):
    lista_ordenada = sorted(lista)
    len_lista = len(lista)
    
    quartis = []
    if len_lista % 2 == 0:
        posicao_quartis = []
        
        for quartil in range(1, 4):
            posicao = (quartil * (len_lista + 1)) / 4
            inteiro_posicao = int(posicao)
            decimal_posicao = posicao - inteiro_posicao
            
            diferenca = lista_ordenada[inteiro_posicao] - lista_ordenada[inteiro_posicao - 1]
            
            proporcao = diferenca * decimal_posicao
            
            quartis.append(lista_ordenada[inteiro_posicao - 1] + proporcao)
            
    else:
        for quartil in range(1, 4):
            posicao = int((quartil * (len_lista + 1)) / 4)
            quartis.append(lista_ordenada[posicao-1])
            
    return quartis

## Exercício 1: Qual equipe teve melhor desempenho?

In [67]:
equipe_a = [42, 38, 45, 41, 39, 44, 38]
equipe_b = [30, 52, 35, 48, 29, 55, 38]

print(f"{'=' * 30} Dados das Equipes  {'=' * 30}")

print(f"Média da equipe A: {statistics.mean(equipe_a)}")
print(f"Média da equipe B: {statistics.mean(equipe_b)}")

print(f"{'=' * 80}")

print(f"Mediana da equipe A: {statistics.median(equipe_a)}")
print(f"Mediana da equipe B: {statistics.median(equipe_b)}")

print(f"{'=' * 80}")

print(f"Moda da equipe A: {calcular_moda(equipe_a)}")
print(f"Moda da equipe B: {calcular_moda(equipe_b)}")

print(f"{'=' * 80}")

print(f"Amplitude da equipe A: {calcula_amplitude(equipe_a)}")
print(f"Amplitude da equipe B: {calcula_amplitude(equipe_b)}")

============================== Dados das Equipes  ==============================
Média da equipe A: 41
Média da equipe B: 41
Mediana da equipe A: 41
Mediana da equipe B: 38
Moda da equipe A: ('Unimodal', [38])
Moda da equipe B: Amodal
Amplitude da equipe A: 7
Amplitude da equipe B: 26


### Conclusões:

As duas equipes possuem média de **41 chamados**

## Exercício 2: Qual máquina é mais regular?

In [68]:
maquina_a = [98, 101, 100, 99, 102, 100, 101]
maquina_b = [82, 115, 91, 108, 120, 86, 99]

print(f"{'=' * 30} Dados das máquinas  {'=' * 30}")

print(f"Média da máquina A: {statistics.mean(maquina_a):.2f}")
print(f"Média da máquina B: {statistics.mean(maquina_b):.2f}")

print(f"{'=' * 80}")

print(f"Variância populacional da máquina A: {statistics.pvariance(maquina_a):.2f}")
print(f"Variância populacional da máquina B: {statistics.pvariance(maquina_b):.2f}")

print(f"{'=' * 80}")

print(f"Desvio padrão populacional da máquina A: {statistics.pstdev(maquina_a):.2f}")
print(f"Desvio padrão populacional da máquina B: {statistics.pstdev(maquina_b):.2f}")

print(f"{'=' * 80}")

print(f"Amplitude da máquina A: {calcula_amplitude(maquina_a)}")
print(f"Amplitude da máquina B: {calcula_amplitude(maquina_b)}")

print(f"{'=' * 80}")

print(f"Coeficiente de variação da máquina A: {calcula_coeficiente_variacao(maquina_a):.2f}%")
print(f"Coeficiente de variação da máquina B: {calcula_coeficiente_variacao(maquina_b):.2f}%")

============================== Dados das máquinas  ==============================
Média da máquina A: 100.14
Média da máquina B: 100.14
Variância populacional da máquina A: 1.55
Variância populacional da máquina B: 184.41
Desvio padrão populacional da máquina A: 1.25
Desvio padrão populacional da máquina B: 13.58
Amplitude da máquina A: 4
Amplitude da máquina B: 38
Coeficiente de variação da máquina A: 1.34%
Coeficiente de variação da máquina B: 14.65%


### Conclusões:

Ainda que a média seja a mesma para ambas as máquinas, a máquina A é mais regular, pois apresenta menor coeficiente de variação (1.34%) em comparação com a máquina B (14.65%). Isso indica que os valores da máquina A estão mais próximos da média, ou seja, a máquina operou de forma mais consistente, como também é evidenciadp pela amplitude de 4 da máquina A, que é significativamente menor do que a amplitude de 38 da máquina B.

## Exercício 3: Análise dos tempos de entrega?

In [69]:
tempos = [48,32,55,41,37,62,29,45,51,35,70]

print(f"{'=' * 30} Dados dos tempos de entrega  {'=' * 30}")

print(f"Média dos tempos de entrega: {statistics.mean(tempos):.2f}")

print(f"{'=' * 90}")

print(f"Mediana dos tempos de entrega: {statistics.median(tempos)}")

print(f"{'=' * 90}")

print(f"Quartis dos tempos de entrega: ")
quartis = calcular_quartis(tempos)
for i in range(1, 4):
    print(f"{i}º Quartil: {quartis[i - 1]}")

print(f"{'=' * 90}")

print(f"Amplitude dos tempos de entrega: {calcula_amplitude(tempos)}")
print(f"Variância populacional dos tempos de entrega: {statistics.pvariance(tempos):.2f}")
print(f"Desvio padrão populacional dos tempos de entrega: {statistics.pstdev(tempos):.2f}")
print(f"Coeficiente de variação dos tempos de entrega: {calcula_coeficiente_variacao(tempos):.2f}%")

print(f"{'=' * 90}")

print(f"Intervalo interquartil dos tempos de entrega: {quartis[2] - quartis[0]}")

============================== Dados dos tempos de entrega  ==============================
Média dos tempos de entrega: 45.91
Mediana dos tempos de entrega: 45
Quartis dos tempos de entrega: 
1º Quartil: 35
2º Quartil: 45
3º Quartil: 55
Amplitude dos tempos de entrega: 41
Variância populacional dos tempos de entrega: 150.45
Desvio padrão populacional dos tempos de entrega: 12.27
Coeficiente de variação dos tempos de entrega: 28.02%
Intervalo interquartil dos tempos de entrega: 20


### Conclusões:

- Um gestor afirma: **"A maioria das entregas ocorre muito próxima da média."** Os resultados encontrados sustentam essa afirmação?

Os resultados encontrados não sustentam a afirmação do gestor. O coeficiente de variação dos tempos de entrega é de 28.02%, o que indica uma variabilidade significativa em relação à média. Além disso, a amplitude dos tempos de entrega é de 41, mostrando que há uma diferença considerável entre os valores mais baixos e mais altos. O intervalo interquartil de 20 também sugere que os dados estão dispersos, com uma diferença substancial entre o primeiro e o terceiro quartil. Portanto, a maioria das entregas não ocorre muito próxima da média, indicando que há uma variabilidade considerável nos tempos de entrega.

## Prática

In [70]:
from datetime import datetime

data_pedido = "12/09/2026"
data_entrega = "18/09/2026"

data_pedido = datetime.strptime(data_pedido, "%d/%m/%Y")
data_entrega = datetime.strptime(data_entrega, "%d/%m/%Y")

dias_entrega = (data_entrega - data_pedido).days

print(f"Data pedido formatada: {data_pedido.strftime('%d/%m/%Y')}")
print(f"Data entrega formatada: {data_entrega.strftime('%d/%m/%Y')}")

print(f"Dias entre o pedido e a entrega: {dias_entrega} dias")

if dias_entrega <= 5:
    print("A entrga foi realizada dentro de 5 dias úteis")
else:
    print("A entrega demorou mais de 5 dias úteis")

Data pedido formatada: 12/09/2026
Data entrega formatada: 18/09/2026
Dias entre o pedido e a entrega: 6 dias
A entrega demorou mais de 5 dias úteis


## Prática

In [71]:
vendedores = [
    {"nome": "Ana", "vendas": 12500},
    {"nome": "Carlos", "vendas": 9800},
    {"nome": "Marina", "vendas": 15700},
    {"nome": "Rafael", "vendas": 11300}
]

for numero, vendedor in enumerate(vendedores, start=1):
    if vendedor['vendas'] >= 10000:
        print(f"Vendedor {numero}: {vendedor['nome']} - Vendas: R${vendedor['vendas']:.2f} | Meta atingida")
    else:
        print(f"Vendedor {numero}: {vendedor['nome']} - Vendas: R${vendedor['vendas']:.2f} | Meta não atingida")

Vendedor 1: Ana - Vendas: R$12500.00 | Meta atingida
Vendedor 2: Carlos - Vendas: R$9800.00 | Meta não atingida
Vendedor 3: Marina - Vendas: R$15700.00 | Meta atingida
Vendedor 4: Rafael - Vendas: R$11300.00 | Meta atingida
